src/MUSI6106Exec/MUSI6106Exec.cpp

```diff
@@ -5,95 +5,172 @@
 #include "MUSI6106Config.h"
 
 #include "AudioFileIf.h"
+#include "CombFilterIf.h"
+#include "RingBuffer.h"
+#include "Vibrato.h"
+
+#define M_PIl          3.141592653589793238462643383279502884L /* pi */
 
 using std::cout;
 using std::endl;
 
 // local function declarations
 void    showClInfo ();
 
-/////////////////////////////////////////////////////////////////////////////////
-// main function
-int main(int argc, char* argv[])
-{
-    std::string             sInputFilePath,                 //!< file paths
-                            sOutputFilePath;
+typedef struct CombFilterArgs {
+    CCombFilterIf::CombFilterType_t filterType = CCombFilterIf::CombFilterType_t::kCombFIR;
+    float delay = 0.1F;
+    float gain = 0.5F;
+    std::string inputPath = "fake_id.wav";
+    std::string outputPath = "out.wav";
+} CombFilterArgs_t;
+
+typedef struct VibratoArgs {
+    CCombFilterIf::CombFilterType_t filterType = CCombFilterIf::CombFilterType_t::kCombFIR;
+    float vibratoWidth = 0.01F;
+    float vibratoFrequency = 6.6F;
+    std::string inputPath = "sin.wav";
+    std::string outputPath = "out.wav";
+} VibratoArgs_t;
+
+void printUsage() {
+    std::printf("Usage: MUSI6106Exec.exe [option]\n\n");
+    std::printf("Options and arguments:\n");
+    std::printf("-w: The vibrato width in Hz.\n");
+    std::printf("    -w <width>: set the width of vibrato effector in Hz. E.g. for 440 sinusoid & 10 Hz vibrato\n");
+    std::printf("                width, it will oscillate in 430-450Hz.\n");
+    std::printf("-f: The vibrato frequency in Hz.\n");
+    std::printf("    -f <freq>: set the frequency of vibrato effector in Hz. E.g. Frequency 6Hz -> vibrato 6 times/s.\n");
+    std::printf("-i: The path to input wave file.\n");
+    std::printf("    -i <path>: load the wave file of <path>. Notice you need to include filename.\n");
+    std::printf("-o: The path to output wave file.\n");
+    std::printf("    -o <path>: write the filtered signal into <path>. Notice you need to include filename\n");
+    std::printf("-h --help: print this help message.\n");
+    std::printf("Examples:\n");
+    std::printf("- Read 'fake_id.wav' and apply vibrato effector, with vibrato width=10Hz, frequency=6.6Hz,\n");
+    std::printf("  write the filtered signal into 'out.wav':\n");
+    std::printf("  `MUSI6106Exec.exe -w 10 -f 6.6 -i fake_id.wav -o out.wav`\n");
+    std::printf("- Display this help:\n");
+    std::printf("  `MUSI6106Exec.exe -h`\n");
+    std::printf("  or\n");
+    std::printf("  `MUSI6106Exec.exe --help`\n");
+}
 
-    static const int        kBlockSize = 1024;
+/*
+Simple command line argument parser.
+*/
+VibratoArgs_t parseArg(int argc, char** argv) {
+    VibratoArgs_t result;
+    for (int i = 0; i < argc; ++i) {
+        if (argv[i][0] == '-') { /* this is an argument */
+            if (argv[i][1] == 'w') {
+                if (i == argc - 1) throw(std::exception("Incomplete argument of delay!"));
```

Comment:
The std::exception with a message seems to be a Microsoft Visual Studio extension. I am not sure if the standard library allows the exception message to be displayed and therefore, I faced some issues while building the code. 
Using std::runtime_error("message") is another option.